lib/internal/modules/esm/loader.js

```diff
@@ -797,41 +815,48 @@ class ESMLoader {
       );
     }
     const chain = this.#resolvers;
+    const context = {
+      conditions: DEFAULT_CONDITIONS,
+      importAssertions,
+      parentURL,
+    };
     const meta = {
       chainFinished: null,
+      context,
       hookErrIdentifier: '',
       hookIndex: chain.length - 1,
       hookName: 'resolve',
       shortCircuited: false,
     };
 
-    const context = {
-      conditions: DEFAULT_CONDITIONS,
-      importAssertions,
-      parentURL,
-    };
-    const validate = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
-
+    const validateArgs = (hookErrIdentifier, { 0: suppliedSpecifier, 1: ctx }) => {
       validateString(
         suppliedSpecifier,
         `${hookErrIdentifier} specifier`,
       ); // non-strings can be coerced to a url string
 
-      validateObject(ctx, `${hookErrIdentifier} context`);
+      if (ctx) validateObject(ctx, `${hookErrIdentifier} context`);
+    };
+    const validateOutput = (hookErrIdentifier, output) => {
+      if (
+        typeof output !== 'object' || // [2]
+        output === null ||
+        (output.url == null && typeof output.then === 'function')
```

Comment:
Resolve is still async, isn't it?
```suggestion
        output === null
```

Reply:
Gah. Copy+pasta 🤦‍♂️ thanks for catching